### ISSUE TITLE
Heuristic scheduler chooses disabled framework

### ISSUE THREAD

**Issue by ms705** [camsas/Musketeer#6]
I'm running the following BEER query, with only Hadoop enabled (`--use_frameworks='hadoop'`):

```
CREATE RELATION edges WITH COLUMNS (INTEGER, INTEGER),
SELECT [edges_0, edges_1] FROM (edges) WHERE [(edges_1 < 5)] AS edges_sel,
AGG [edges_sel_1, +] FROM (edges_sel) GROUP BY [edges_sel_0] AS sum_group_by,
AGG [sum_group_by_1, +] FROM (sum_group_by) GROUP BY [sum_group_by_0] AS sec_sum_group_by
```

... and get the following output from the heuristic scheduler (`--use_heuristic=true`):

```
$ build/musketeer --dry_run --run_daemon=0 --beer_query=tests/foo.rap --root_dir=/tmp/ --output_ir_dag_gv --use_frameworks="hadoop" 
I0706 16:21:06.385500 28560 musketeer.cc:182] Adding Hadoop Framework
I0706 16:21:06.385642 28560 musketeer.cc:261] Looking for new Job to schedule
[...]
digraph OpDAG {
node [shape=box]; edges_sel [label="14"]; 
edges_sel->sum_group_by [label="edges_sel"];
sum_group_by->sec_sum_group_by [label="sum_group_by"];
}
I0706 16:21:06.386896 28560 musketeer.cc:305] Scheduling entire DAG
I0706 16:21:06.386917 28560 scheduler_dynamic.cc:170] Determine inputs size for DAG
I0706 16:21:06.386996 28560 scheduler_dynamic.cc:195] Size of: edges is: 0
I0706 16:21:06.387027 28560 scheduler_dynamic.cc:240] DynamicSchedule DAG
I0706 16:21:06.387049 28560 scheduler_dynamic.cc:439] Topological order: edges_sel
I0706 16:21:06.387071 28560 scheduler_dynamic.cc:439] Topological order: sum_group_by
I0706 16:21:06.387092 28560 scheduler_dynamic.cc:439] Topological order: sec_sum_group_by
I0706 16:21:06.387111 28560 utils.cc:256] Node order after optimisation: edges_sel
I0706 16:21:06.387125 28560 utils.cc:256] Node order after optimisation: sum_group_by
I0706 16:21:06.387138 28560 utils.cc:256] Node order after optimisation: sec_sum_group_by
I0706 16:21:06.387158 28560 scheduler_dynamic.cc:342] Refresh rel size of edges_sel is 0
I0706 16:21:06.387179 28560 scheduler_dynamic.cc:342] Refresh rel size of sum_group_by is 0
I0706 16:21:06.387199 28560 scheduler_dynamic.cc:342] Refresh rel size of sec_sum_group_by is 0
I0706 16:21:06.387222 28560 scheduler_dynamic.cc:602] ComputeHeuristic
I0706 16:21:06.387487 28560 scheduler_dynamic.cc:648] 1 1 20 0
I0706 16:21:06.387503 28560 scheduler_dynamic.cc:648] 2 1 20 0
I0706 16:21:06.387511 28560 scheduler_dynamic.cc:648] 2 2 40 1
I0706 16:21:06.387517 28560 scheduler_dynamic.cc:648] 3 1 4294967295 0
I0706 16:21:06.387524 28560 scheduler_dynamic.cc:648] 3 2 40 2
I0706 16:21:06.387531 28560 scheduler_dynamic.cc:648] 3 3 60 2
I0706 16:21:06.387537 28560 scheduler_dynamic.cc:671] Schedulable operators: [1, 3]
SCHEDULER TIME: 0.000333
I0706 16:21:06.387599 28560 scheduler_dynamic.cc:300] Dispatching relation sum_group_by in framework hadoop
[...]
I0706 16:21:07.162153 28560 scheduler_dynamic.cc:170] Determine inputs size for DAG
I0706 16:21:07.162215 28560 scheduler_dynamic.cc:195] Size of: edges is: 0
I0706 16:21:07.162271 28560 scheduler_dynamic.cc:759] Size of output: sum_group_by is: 0
I0706 16:21:07.162322 28560 scheduler_dynamic.cc:283] Running operators 1 2 on hadoop
I0706 16:21:07.162343 28560 scheduler_dynamic.cc:289] Number of operators scheduled: 2
I0706 16:21:07.162372 28560 scheduler_dynamic.cc:342] Refresh rel size of sec_sum_group_by is 0
I0706 16:21:07.162390 28560 scheduler_dynamic.cc:602] ComputeHeuristic
I0706 16:21:07.162441 28560 scheduler_dynamic.cc:648] 1 1 20 0
I0706 16:21:07.162463 28560 scheduler_dynamic.cc:671] Schedulable operators: [1, 1]
SCHEDULER TIME: 9.2e-05
I0706 16:21:07.162528 28560 scheduler_dynamic.cc:300] Dispatching relation sec_sum_group_by in framework spark
Segmentation fault
```

... which is strange, since Spark should not even be an option.

Backtrace:

```
#0  0x0000000000879de0 in ?? ()
#1  0x0000000000594057 in musketeer::scheduling::SchedulerDynamic::DispatchWithHistory (this=0x837570, bind=..., 
    nodes=std::vector of length 1, capacity 1 = {...}, relation="sec_sum_group_by")
    at musketeer/src/scheduling/scheduler_dynamic.cc:305
#2  0x00000000005948f7 in musketeer::scheduling::SchedulerDynamic::DynamicScheduleDAG (this=0x837570, 
    dag=std::vector of length 1, capacity 1 = {...})
    at musketeer/src/scheduling/scheduler_dynamic.cc:281
#3  0x00000000004174c3 in main (argc=7, argv=0x7fffffffe3f8) at musketeer/src/musketeer.cc:307
```

The exhaustive scheduler gets it right:

```
$ build/musketeer --dry_run --run_daemon=0 --beer_query=tests/foo.rap --root_dir=/tmp/ --output_ir_dag_gv --use_frameworks="hadoop" --use_heuristic=false
I0706 16:27:10.152679 28792 musketeer.cc:182] Adding Hadoop Framework
I0706 16:27:10.152823 28792 musketeer.cc:261] Looking for new Job to schedule
[...]
digraph OpDAG {
node [shape=box]; edges_sel [label="14"]; 
edges_sel->sum_group_by [label="edges_sel"];
sum_group_by->sec_sum_group_by [label="sum_group_by"];
}
I0706 16:27:10.154114 28792 musketeer.cc:305] Scheduling entire DAG
I0706 16:27:10.154135 28792 scheduler_dynamic.cc:170] Determine inputs size for DAG
I0706 16:27:10.154217 28792 scheduler_dynamic.cc:195] Size of: edges is: 0
I0706 16:27:10.154248 28792 scheduler_dynamic.cc:240] DynamicSchedule DAG
I0706 16:27:10.154271 28792 scheduler_dynamic.cc:439] Topological order: edges_sel
I0706 16:27:10.154294 28792 scheduler_dynamic.cc:439] Topological order: sum_group_by
I0706 16:27:10.154316 28792 scheduler_dynamic.cc:439] Topological order: sec_sum_group_by
I0706 16:27:10.154336 28792 utils.cc:256] Node order after optimisation: edges_sel
I0706 16:27:10.154351 28792 utils.cc:256] Node order after optimisation: sum_group_by
I0706 16:27:10.154366 28792 utils.cc:256] Node order after optimisation: sec_sum_group_by
I0706 16:27:10.154387 28792 scheduler_dynamic.cc:342] Refresh rel size of edges_sel is 0
I0706 16:27:10.154408 28792 scheduler_dynamic.cc:342] Refresh rel size of sum_group_by is 0
I0706 16:27:10.154429 28792 scheduler_dynamic.cc:342] Refresh rel size of sec_sum_group_by is 0
I0706 16:27:10.177091 28792 scheduler_dynamic.cc:558] The minimum cost of running the DAG: 40
I0706 16:27:10.177129 28792 scheduler_dynamic.cc:562] Cur cost: 40
I0706 16:27:10.177139 28792 scheduler_dynamic.cc:570] ---------- Job boundary ----------
I0706 16:27:10.177146 28792 scheduler_dynamic.cc:574] sec_sum_group_by
I0706 16:27:10.177157 28792 scheduler_dynamic.cc:562] Cur cost: 20
I0706 16:27:10.177165 28792 scheduler_dynamic.cc:570] ---------- Job boundary ----------
I0706 16:27:10.177172 28792 scheduler_dynamic.cc:574] edges_sel
I0706 16:27:10.177181 28792 scheduler_dynamic.cc:574] sum_group_by
SCHEDULER TIME: 0.029416
I0706 16:27:10.188161 28792 scheduler_dynamic.cc:300] Dispatching relation sum_group_by in framework hadoop
[...]
I0706 16:27:10.936656 28792 scheduler_dynamic.cc:170] Determine inputs size for DAG
I0706 16:27:10.936723 28792 scheduler_dynamic.cc:195] Size of: edges is: 0
I0706 16:27:10.936769 28792 scheduler_dynamic.cc:759] Size of output: sum_group_by is: 0
I0706 16:27:10.936813 28792 scheduler_dynamic.cc:283] Running operators 1 2 on hadoop
I0706 16:27:10.936827 28792 scheduler_dynamic.cc:289] Number of operators scheduled: 2
I0706 16:27:10.936861 28792 scheduler_dynamic.cc:342] Refresh rel size of sec_sum_group_by is 0
I0706 16:27:10.957470 28792 scheduler_dynamic.cc:558] The minimum cost of running the DAG: 20
I0706 16:27:10.957501 28792 scheduler_dynamic.cc:562] Cur cost: 20
I0706 16:27:10.957509 28792 scheduler_dynamic.cc:570] ---------- Job boundary ----------
I0706 16:27:10.957517 28792 scheduler_dynamic.cc:574] sec_sum_group_by
SCHEDULER TIME: 0.029033
I0706 16:27:10.965978 28792 scheduler_dynamic.cc:300] Dispatching relation sec_sum_group_by in framework hadoop
[...]
I0706 16:27:11.746538 28792 scheduler_dynamic.cc:170] Determine inputs size for DAG
I0706 16:27:11.746582 28792 scheduler_dynamic.cc:759] Size of output: sec_sum_group_by is: 0
I0706 16:27:11.746605 28792 scheduler_dynamic.cc:283] Running operators 3 3 on hadoop
I0706 16:27:11.746618 28792 scheduler_dynamic.cc:289] Number of operators scheduled: 1
I0706 16:27:11.746676 28792 musketeer.cc:329] Finished scheduling job
```

**Comment by ICGog** [comment]
This issue has been addressed in 09242163bca00fbd9d7ec8823a1d6b420b0b14b9.

**Comment by ms705** [comment]
Confirmed that this fixes the problem with the above workflow.
